### ISSUE TITLE
Investigate blends in SDSS that we cannot deblend

### ISSUE THREAD

**Issue by ismael-mendoza** [prob-ml/bliss#461]
I've identified three blends in SDSS that we cannot deblend. We should understand why for the paper I think, at least in terms of limitations even if it turns out the issue is deeper than a simple bug so that we might have to delay solving it.

<img width="766" alt="Screen Shot 2022-02-27 at 4 47 51 PM" src="https://user-images.githubusercontent.com/11745764/155902671-b816da69-f4f4-4bf9-8d3d-be8d137d9c70.png">
<img width="766" alt="Screen Shot 2022-02-27 at 4 47 51 PM" src="https://user-images.githubusercontent.com/11745764/155902674-5ccc29f9-6dfd-4bcf-9b24-ca544d12e561.png">
<img width="822" alt="Screen Shot 2022-02-27 at 4 59 58 PM" src="https://user-images.githubusercontent.com/11745764/155902679-f3b36eb6-64c0-474d-8af9-c32018b75c9f.png">

**Comment by ismael-mendoza** [comment]
Here is one that we can 

<img width="784" alt="image" src="https://user-images.githubusercontent.com/11745764/155902747-742bf90f-fa76-4d04-a91d-430571251ee1.png">

**Comment by jeff-regier** [comment]
Does PR #561 solve some/most of these problems?

**Comment by ismael-mendoza** [comment]
I will retrain and investigate this soon

**Comment by dereklhansen** [comment]
On my end, it doesn't look like BLISS gets these blends after #561.

For the first, BLISS and PHOTO make the same prediction (one galaxy in the center).
For the second, BLISS assigns a 34% detection probability and PHOTO predicts it.
For the third, BLISS assigns a 0% probability and PHOTO does not predict it.

So really only the second blend is something we should maybe be getting, and the good news is that the probability is pretty high

**Comment by ismael-mendoza** [comment]
Thanks @dereklhansen - on my end I find that we are still not getting some of the easy blends. I guess I can take a look at PHOTO and see what it says on the frame but these look like pretty easy to detect. 

![image](https://user-images.githubusercontent.com/11745764/165403646-1d0bdf6e-e7e9-410d-91b3-d2adafd8bb43.png)

![image](https://user-images.githubusercontent.com/11745764/165403803-e7a3da3f-5320-4ae1-a892-1cddd77c7074.png)

**Comment by jeff-regier** [comment]
> For the second, BLISS assigns a 34% detection probability and PHOTO predicts it.

I took a look at the second one (with coordinates 336.7208, -1.0159) in images from the Hyper Suprime-Cam, which is more sensitive that SDSS, and I think I see why it's difficult:
<img width="201" alt="Screen Shot 2022-04-26 at 8 36 31 PM" src="https://user-images.githubusercontent.com/8953355/165415365-c359a247-ab8a-4685-a060-f1638bf17f2c.png">

It's a spiral galaxy with highly visible arms. Our bulge-plus-disk galaxy model can't represent arms. Usually this doesn't matter because arms either nonexistent, or they are faint and get blurred by the PSF. This galaxies is a weird one though. I don't think bliss misses it because of blending.

**Comment by jeff-regier** [comment]
> For the first, BLISS and PHOTO make the same prediction (one galaxy in the center).

Here's the first one in DeCALS:
<img width="109" alt="Screen Shot 2022-04-26 at 8 51 45 PM" src="https://user-images.githubusercontent.com/8953355/165416579-24ff9a52-1eb5-487e-8a72-8ddc055e766d.png">

It's clearly two objects, but it's pretty hard to guess that from just one SDSS band. However, I do think bliss should say the probability of two objects is nonzero (and probably > 5%). If it doesn't do that, I wonder if our latest training procedure isn't using enough blends. (In our model, what are the odds of generating two sources this close? How many sources this blended do we expect to generate during training?)

One idea: For training, greatly increase the density of light sources (like we had been doing, back when bliss wasn't work as well), but up/down weight the training examples based on how many light sources are present. Maybe this would give the encoder more exposure to blends (while still being consistent with the prior).

**Comment by jeff-regier** [comment]
> For the third, BLISS assigns a 0% probability and PHOTO does not predict it.

The third one may be more like 7 different objects, with many very faint. Here it is in HSC, with the DeCALS catalog overlayed:
<img width="273" alt="Screen Shot 2022-04-26 at 9 03 59 PM" src="https://user-images.githubusercontent.com/8953355/165417889-a838fd82-9c40-4f31-a9f5-ed0073d1c507.png">

The DeCALS catalog isn't necessarily right here. But it makes the point that there's more than meets the eye in the SDSS frame. Given that, it seems OK to estimate 0% for this one because the faint/undetectable objects make it so that the detectible objects look like nothing our model could generate. If we want to get this one right, we'd need to have our model generate faint/undetectable objects too, so that the encoder is robust to this type of variation.

**Comment by ismael-mendoza** [comment]
> So the VAE is getting these "easy" blends right, but the autoencoder is not? If so, that may be unrelated to the three difficult blends in the original post.

I think the VAE and AE are still getting (approximately) the same results. Sorry for the confusing wording. The second blend I posted was not included in the original three but I would still consider it "easy". 


> I took a look at the second one (with coordinates 336.7208, -1.0159) in images from the Hyper Suprime-Cam, which is more sensitive that SDSS, and I think I see why it's difficult:

Yeah, I don't think the spiral arms are visible thanks to the blurring of the PSF in SDSS. I think we miss the left one though, which should be easier in any case.

**Comment by jeff-regier** [comment]
What are the coordinates for the second blend you posted?

**Comment by jeff-regier** [comment]
> Yeah, I don't think the spiral arms are visible thanks to the blurring of the PSF in SDSS. I think we miss the left one though, which should be easier in any case.

I see. The arms of the right object may be interfering with our detection of the left one. What was the probability of detection for the left one?

**Comment by dereklhansen** [comment]
My version of the code gets Ismael's second blend just fine:

![image](https://user-images.githubusercontent.com/12701492/165550850-69fc5308-56ee-4219-b639-c7a28af42f54.png)

Of the two objects, their probabilites are (reading left-to-right) 0.9937 and 0.8167. This matches what Jeff is saying, since the right galaxy has a shape that we are not generating.

**Comment by dereklhansen** [comment]
I wonder if there are issues with the training procedure for LocationEncoder. Currently, we run the training procedure and afterwards pick the weights with the lowest validation loss. This differs from my understanding of how early stopping usually works, where training is halted if if we haven't decreased validation loss in X number of epochs. 

Perhaps the bigger issue is that the validation dataset is normally fixed in these kind of schemes, but we randomly generate it each time. This adds more noise to the process and prevents an apples-to-apples comparison. Specifically, we could get a lower validation loss at a particular epoch because the validation set got easier, not necessarily because the model improved. And since we're taking the minimum, this noise does make a big difference, especially near the end of training and the improvements in the model are dwarfed by the noise in the sampling process.

I only bring this up because, on my case studiy, I changed the training process for the LocationEncoder when I started using Galsim directly to have less epochs but more steps per epoch. This was motivated by computation, since it allowed parallel workers to generate Galsim images. However, this had a side-effect of increasing the size of the batch used for validation, which reduced the aforementioned noise in the validation set I mentioned. This could be why my results seem to be better here.

**Comment by jeff-regier** [comment]
> Perhaps the bigger issue is that the validation dataset is normally fixed in these kind of schemes, but we randomly generate it each time.

Good point!
